engines/freescape/freescape.cpp

```diff
@@ -0,0 +1,609 @@
+				move(kForwardMovement, _scaleVector.x(), deltaTime);
+			else if (event.kbd.keycode == Common::KEYCODE_k || event.kbd.keycode == Common::KEYCODE_DOWN)
+				move(kBackwardMovement, _scaleVector.x(), deltaTime);
+			else if (event.kbd.keycode == Common::KEYCODE_LEFT)
+				move(kLeftMovement, _scaleVector.y(), deltaTime);
+			else if (event.kbd.keycode == Common::KEYCODE_RIGHT)
+				move(kRightMovement, _scaleVector.y(), deltaTime);
+			else if (event.kbd.keycode == Common::KEYCODE_KP5 || event.kbd.keycode == Common::KEYCODE_KP0)
+				shoot();
+			else if (event.kbd.keycode == Common::KEYCODE_p)
+				rotate(0, 5);
+			else if (event.kbd.keycode == Common::KEYCODE_l)
+				rotate(0, -5);
+			else if (event.kbd.keycode == Common::KEYCODE_u)
+				rotate(180, 0);
+			else if (event.kbd.keycode == Common::KEYCODE_q)
+				rotate(-_angleRotations[_angleRotationIndex], 0);
+			else if (event.kbd.keycode == Common::KEYCODE_w)
+				rotate(_angleRotations[_angleRotationIndex], 0);
+			else if (event.kbd.keycode == Common::KEYCODE_r)
+				rise();
+			else if (event.kbd.keycode == Common::KEYCODE_f)
+				lower();
+			else if (event.kbd.keycode == Common::KEYCODE_n) {
+				_noClipMode = !_noClipMode;
+				_flyMode = _noClipMode;
+			} else if (event.kbd.keycode == Common::KEYCODE_ESCAPE)
+				openMainMenuDialog();
+			else if (event.kbd.keycode == Common::KEYCODE_SPACE) {
+				_shootMode = !_shootMode;
+				if (!_shootMode) {
+					_crossairPosition.x = _screenW / 2;
+					_crossairPosition.y = _screenH / 2;
+				}
+			} else
+				pressedKey(event.kbd.keycode);
+			break;
+
+		case Common::EVENT_QUIT:
+		case Common::EVENT_RETURN_TO_LAUNCHER:
+			quitGame();
+			return;
+			break;
+
+		case Common::EVENT_MOUSEMOVE:
+			mousePos = event.mouse;
+
+			if (_demoMode)
+				g_system->warpMouse(mousePos.x, mousePos.y);
+
+			if (_shootMode) {
+				_crossairPosition = mousePos;
+				break;
+			}
+
+			if (mousePos.x <= 5 || mousePos.x >= _screenW - 5) {
+				g_system->warpMouse(_screenW / 2, mousePos.y);
+
+				_lastMousePos.x = _screenW / 2;
+				_lastMousePos.y = mousePos.y;
+				if (mousePos.x <= 5)
+					mousePos.x = _lastMousePos.x + 3;
+				else
+					mousePos.x = _lastMousePos.x - 3;
+
+				mousePos.y = _lastMousePos.y;
+
+			} else if (mousePos.y <= 5 || mousePos.y >= _screenH - 5) {
+				g_system->warpMouse(mousePos.x, _screenH / 2);
+				_lastMousePos.x = mousePos.x;
+				_lastMousePos.y = _screenH / 2;
+				if (mousePos.y <= 5)
+					mousePos.y = _lastMousePos.y + 3;
+				else
+					mousePos.y = _lastMousePos.y - 3;
+
+				mousePos.x = _lastMousePos.x;
+			}
+			rotate(_lastMousePos, mousePos);
+			_lastMousePos = mousePos;
+			break;
+
+		case Common::EVENT_LBUTTONDOWN:
+			shoot();
+			break;
+
+		default:
+			break;
+		}
+	}
+}
+
+Common::Error FreescapeEngine::run() {
+	// Initialize graphics
+	_gfx = createRenderer(_system, _screenW, _screenH);
+	_gfx->_isAmiga = isAmiga();
```

Comment:
I think you cannot be both Amiga and Atari. Thus, I'd recommend using as in other engines `getPlatform()` which you implement in the `metaengine.cpp`

Reply:
I just realized that renderer needs to know the render mode actually, not really the platform. This will be refactored soon to use `RenderMode` instead.

Reply:
Fixed